### PR TITLE
Add reasonable application command handling architecture, and comments.

### PR DESCRIPTION
**Pull request by quantatic**
This commit also adds a rustfmt.toml file (though it contains options
which only work on the nightly rustfmt toolchain). We could simplify
the list of format options by a fair bit, but a few very useful options
do still require nightly (`wrap_comments`, for instance). We may want
to consider using a nightly toolchain (perhaps pinned to a specific
release) to run the formatting portion of our build pipeline.

This commit also adds documentation for all implemented features.
Documentation might be a bit rough around the edges, but it should
get the point across, subject to some proof-reading.

### CODE REVIEW

**Comment by quantatic** [comment]
This PR is certainly still a bit rough around the edges, and still subject to quite a bit of change.

File: src/main.rs [hunk 1]
```diff
@@ -1,91 +1,186 @@
...
+    // application commands (in [`APPLICATION_COMMANDS`]). If any match,
+    // we dispatch that application command's handler.
+    //
+    // # Panics
+    //
+    // Panics if none of the available [`APPLICATION_COMMANDS`] match, or if we run
+    // into an error handling this application command.
     async fn interaction_create(&self, ctx: Context, interaction: Interaction) {
-        // Figure out what interaction was sent
+        println!("{:#?}", interaction);
```
Comment (quantatic) [inline src/main.rs:112]: Could probably do with a few more comments around here, and without this `println`.

File: src/main.rs [hunk 2]
```diff
@@ -1,91 +1,186 @@
...
+                            command.data.name
+                        ))
+                    })
                 })
                 .await
+                .expect("failed to send message");
+        }
+    }
+
+    // Dispatched when a guild is created; or an existing guild's data is sent to
```
Comment (quantatic) [inline src/main.rs:138]: Ideally we'll modify this logic to handle global application commands, but I'm using guild-specific registration for more rapid iteration, for now.

File: src/main.rs (hunk 1 above)
Comment (weakphish) [reply src/main.rs:112]: Should we implement formal logging on some level, instead of print statements?

**Comment by weakphish** [comment]
> This commit also adds a rustfmt.toml file (though it contains options which only work on the nightly rustfmt toolchain). We could simplify the list of format options by a fair bit, but a few very useful options do still require nightly (`wrap_comments`, for instance). We may want to consider using a nightly toolchain (perhaps pinned to a specific release) to run the formatting portion of our build pipeline.
> 
> This commit also adds documentation for all implemented features. Documentation might be a bit rough around the edges, but it should get the point across, subject to some proof-reading.

I'm thinking that we maybe stay away from dependency on nightly. That can be offputting to potential contributors and isn't necessarily required.

File: src/main.rs [hunk 3]
```diff
@@ -1,91 +1,186 @@
...
+/// [`ApplicationCommandInteraction`].
+pub const APPLICATION_COMMANDS: &[&(dyn ApplicationCommandHandler + Sync)] =
+    &[&PingApplicationCommand];
+
+/// This trait represents a handler for an [`ApplicationCommandInteraction`].
+/// Unit structs implementing this trait are stored in [`APPLICATION_COMMANDS`],
+/// which are then registered and handled elsewhere.
+///
+/// Methods associated with this trait may find it useful to utilize the passed
+/// [`Context::data`]. This field is extremely useful to persist data across
```
Comment (weakphish) [inline src/main.rs:35]: this is exactly what I was looking for wrt. docs, thank you!

File: src/main.rs [hunk 4]
```diff
@@ -1,91 +1,186 @@
...
         if let Interaction::ApplicationCommand(command) = interaction {
-            let content = match command.data.name.as_str() {
-                "ping" => "Hey!".to_string(),
-                _ => "I didn't get that.".to_string(),
-            };
+            for application_command in APPLICATION_COMMANDS {
+                if application_command.get_name() == command.data.name {
+                    application_command
+                        .handle_command(&ctx, command)
+                        .await
```
Comment (weakphish) [inline src/main.rs:118]: this looks super clean

File: src/main.rs (hunk 2 above)
Comment (weakphish) [reply src/main.rs:138]: that's fine for now - I think only @tcheinen and @ProfDoof were planning on maybe expanding this, so for our current scope that's fine

File: src/main.rs [hunk 5]
```diff
@@ -1,91 +1,186 @@
...
-
-    // The Application Id is usually the Bot User Id.
-    let application_id: u64 = fs::read_to_string("app_id")
-        .expect("Could not read file.")
+async fn main() -> Result<(), Box<dyn Error>> {
+    let application_id: u64 = env::var("APPLICATION_ID")
+        .map_err(|_| "Expected APPLICATION_ID in the environment")?
         .parse()
-        .expect("Something went wrong reading the file");
+        .map_err(|_| "APPLICATION_ID is not a valid id")?;
```
Comment (weakphish) [inline src/main.rs:172]: what's the benefit of a map_err?

File: src/ping.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+/// A basic [`ApplicationCommandHandler`], intended for demonstration purposes.
+///
+/// This [`ApplicationCommandHandler`] simply replies to `/ping` with
+/// `Pong, <user>!`.
+pub struct PingApplicationCommand;
+
+#[async_trait]
+impl ApplicationCommandHandler for PingApplicationCommand {
+    async fn register_command(
```
Comment (weakphish) [inline src/ping.rs:33]: is there a potential DRY-violation in re-implementing `register_command` for every command?

**Review by weakphish — COMMENTED** [review]
some questions, but otherwise LGTM!

**Comment by ProfDoof** [comment]
So, my thought is that if we're just doing documentation verification then there should be a way to do that step without doing actual compilation. And if that's the case then having the nightly build used for documentation verification is no problem because it's not actually being used for compilation. However, I don't actually know that that's a thing.

**Comment by quantatic** [comment]
> > This commit also adds a rustfmt.toml file (though it contains options which only work on the nightly rustfmt toolchain). We could simplify the list of format options by a fair bit, but a few very useful options do still require nightly (`wrap_comments`, for instance). We may want to consider using a nightly toolchain (perhaps pinned to a specific release) to run the formatting portion of our build pipeline.
> > This commit also adds documentation for all implemented features. Documentation might be a bit rough around the edges, but it should get the point across, subject to some proof-reading.
> 
> I'm thinking that we maybe stay away from dependency on nightly. That can be offputting to potential contributors and isn't necessarily required.

Unfortunately, taking a look at the [rustfmt configuration options](https://rust-lang.github.io/rustfmt/?version=v1.4.36), virtually all of these are unstable. As @ProfDoof mentioned, I still intended to actually build/test the core project with a stable release. I plan to run `rustfmt` (using the nightly release) to ensure that formatting is proper, as part of the build pipeline. Thoughts on this? Using `rustup` it's really as simple as:
```bash
$ rustup default nightly(some specific version if we'd like)
$ rustup component add rustfmt # might not be needed actually
$ cargo fmt
$ rustup default stable (or some specific version if we'd like)
```

File: src/main.rs (hunk 5 above)
Comment (quantatic) [reply src/main.rs:172]: I'm rather adverse to panicking in general. `map_err` here gives us a more descriptive error message, while still letting us utilize the "returning `result` from main" pattern.

If you find it more readable, I can just use `.expect` here instead.

File: src/ping.rs (hunk 6 above)
Comment (quantatic) [reply src/ping.rs:33]: How were you thinking of refactoring this logic out? Aside from registering commands with different options, different implementations may store various objects in the global context, in addition to a variety of other checks.

Comment (weakphish) [reply src/ping.rs:33]: That makes sense to me. LGTM

File: src/main.rs (hunk 5 above)
Comment (weakphish) [reply src/main.rs:172]: map_err is fine, I was just curious since I haven't seen this pattern before.

**Review by weakphish — APPROVED** [review]
LGTM. This is a great starting point that will be super handy moving forward in terms of architecture reference.

**Pull request merged**